File: android-app/app/src/main/java/com/lasohealth/android/core/analysis/analyzers/WorkoutEffectivenessAnalyzer.kt
package com.lasohealth.android.core.analysis.analyzers

import com.lasohealth.android.core.analysis.AnalysisContext
import com.lasohealth.android.core.analysis.AnalysisInsight
import com.lasohealth.android.core.analysis.DailySample
import com.lasohealth.android.core.analysis.InsightAnalyzer
import com.lasohealth.android.core.analysis.InsightDirective
import com.lasohealth.android.core.analysis.InsightPriority
import com.lasohealth.android.core.model.HealthMetric
import java.util.Calendar
import java.util.UUID
import kotlin.math.abs

/**
 * Measures workout consistency, VO2 Max response, and calorie efficiency.
 *
 * Generates insights for:
 *   1. Workout consistency score (% of weeks with 3+ sessions over 4 weeks)
 *   2. VO2 Max trend (30-day vs prior 30-day comparison)
 *   3. Calorie efficiency (kcal/min this week vs 30-day average)
 *
 * Ported from iOS WorkoutEffectivenessAnalyzer.swift.
 */
class WorkoutEffectivenessAnalyzer : InsightAnalyzer {

    override val name: String = "WorkoutEffectivenessAnalyzer"

    override suspend fun generateInsights(context: AnalysisContext): List<AnalysisInsight> {
        val insights = mutableListOf<AnalysisInsight>()

        context.timeSeries[HealthMetric.WORKOUT_DURATION]?.let { workoutSamples ->
            analyzeConsistency(workoutSamples)?.let { insights += it }
        }

        context.timeSeries[HealthMetric.VO2_MAX]?.let { vo2Samples ->
            analyzeVO2MaxResponse(vo2Samples)?.let { insights += it }
        }

        val calSamples = context.timeSeries[HealthMetric.ACTIVE_CALORIES]
        val exSamples = context.timeSeries[HealthMetric.EXERCISE_MINUTES]
        if (calSamples != null && exSamples != null) {
            analyzeCalorieEfficiency(calSamples, exSamples)?.let { insights += it }
        }

        return insights
    }

    // ----- 1. Workout Consistency -----

    private fun analyzeConsistency(workoutSamples: List<DailySample>): AnalysisInsight? {
        val now = System.currentTimeMillis()
        val last28 = workoutSamples.filter { it.date > now - 28 * DAY_MS }
        if (last28.size < 7) return null

        // Count workout days per week for the last 4 weeks
        val weekCounts = IntArray(4)
        for (weekOffset in 0 until 4) {
            val weekStart = now - (28 - weekOffset * 7) * DAY_MS
            val weekEnd = now - (21 - weekOffset * 7) * DAY_MS
            weekCounts[weekOffset] = last28.count { it.value > 0 && it.date in weekStart until weekEnd }
        }

        val weeksWithTarget = weekCounts.count { it >= 3 }
        val consistencyScore = weeksWithTarget / 4.0 * 100.0
        val weeklyAvg = weekCounts.sum() / 4.0
        val weekBreakdown = weekCounts.mapIndexed { i, c -> "Week ${i + 1}: $c" }.joinToString(", ")

        return AnalysisInsight(
            id = UUID.randomUUID().toString(),
            title = "Workout Consistency",
            detail = "${consistencyScore.toInt()}% of the last 4 weeks had 3+ workout days " +
                "(avg ${String.format("%.1f", weeklyAvg)}/week). Breakdown: $weekBreakdown.",
            metric = HealthMetric.WORKOUT_DURATION,
            priority = if (consistencyScore < 50) InsightPriority.MEDIUM else InsightPriority.INFORMATIONAL,
            directive = if (consistencyScore < 50) InsightDirective.INCREASE_ACTIVITY else InsightDirective.MAINTAIN,
        )
    }

    // ----- 2. VO2 Max Response -----

    private fun analyzeVO2MaxResponse(vo2Samples: List<DailySample>): AnalysisInsight? {
        val now = System.currentTimeMillis()
        val recent = vo2Samples.filter { it.date > now - 30 * DAY_MS }
        val older = vo2Samples.filter { it.date > now - 60 * DAY_MS && it.date <= now - 30 * DAY_MS }

        if (recent.size < 3 || older.size < 3) return null

        val recentAvg = recent.map { it.value }.average()
        val olderAvg = older.map { it.value }.average()
        if (olderAvg <= 0) return null

        val change = ((recentAvg - olderAvg) / olderAvg) * 100.0

        // Weekly VO2 progression
        val weeklyAvgs = mutableListOf<String>()
        for (weekOffset in 3 downTo 0) {
            val weekStart = now - 7 * (weekOffset + 1).toLong() * DAY_MS
            val weekEnd = now - 7 * weekOffset.toLong() * DAY_MS
            val weekSamples = vo2Samples.filter { it.date in weekStart until weekEnd }
            if (weekSamples.isNotEmpty()) {
                weeklyAvgs += String.format("%.1f", weekSamples.map { it.value }.average())
            }
        }
        val weeklyProgression = if (weeklyAvgs.isNotEmpty()) " Weekly VO2: ${weeklyAvgs.joinToString(" \u2192 ")}." else ""

        return AnalysisInsight(
            id = UUID.randomUUID().toString(),
            title = "VO2 Max Response",
            detail = "Your VO2 Max ${if (change > 0) "improved" else "decreased"} ${String.format("%.1f", abs(change))}% " +
                "over the last 30 days (${String.format("%.1f", olderAvg)} \u2192 ${String.format("%.1f", recentAvg)} " +
                "${HealthMetric.VO2_MAX.unit}).$weeklyProgression",
            metric = HealthMetric.VO2_MAX,
            priority = if (abs(change) > 5) InsightPriority.MEDIUM else InsightPriority.INFORMATIONAL,
            directive = if (change > 0) InsightDirective.MAINTAIN else InsightDirective.INCREASE_ACTIVITY,
        )
    }

    // ----- 3. Calorie Efficiency -----

    private fun analyzeCalorieEfficiency(
        calSamples: List<DailySample>,
        exSamples: List<DailySample>,
    ): AnalysisInsight? {
        val now = System.currentTimeMillis()

        // Build day-aligned pairs
        val calByDay = calSamples.associate { startOfDay(it.date) to it.value }
        val exByDay = exSamples.associate { startOfDay(it.date) to it.value }

        val aligned7d = mutableListOf<Pair<Double, Double>>()
        val aligned30d = mutableListOf<Pair<Double, Double>>()

        for ((day, cal) in calByDay) {
            val ex = exByDay[day] ?: continue
            if (ex <= 0) continue
            if (day > now - 7 * DAY_MS) aligned7d += cal to ex
            if (day > now - 30 * DAY_MS) aligned30d += cal to ex
        }

        if (aligned7d.size < 3 || aligned30d.size < 7) return null

        val efficiency7d = aligned7d.map { it.first / it.second }.average()
        val efficiency30d = aligned30d.map { it.first / it.second }.average()
        if (efficiency30d <= 0) return null

        val change = ((efficiency7d - efficiency30d) / efficiency30d) * 100.0

        return AnalysisInsight(
            id = UUID.randomUUID().toString(),
            title = "Calorie Efficiency",
            detail = "You're burning ${String.format("%.1f", efficiency7d)} kcal/min this week vs " +
                "${String.format("%.1f", efficiency30d)} kcal/min over 30 days " +
                "(${if (change > 0) "+" else ""}${String.format("%.0f", change)}%).",
            metric = HealthMetric.ACTIVE_CALORIES,
            priority = InsightPriority.INFORMATIONAL,
            directive = InsightDirective.INFORMATIONAL,
        )
    }

    // ----- Helpers -----

    private fun startOfDay(epochMs: Long): Long {
        val cal = Calendar.getInstance().apply { timeInMillis = epochMs }
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    companion object {
        private const val DAY_MS = 86_400_000L
    }
}

File: Core/Analysis/WorkoutEffectivenessAnalyzer.swift
import Foundation

/// Measures workout consistency, VO2 Max response, and calorie efficiency
struct WorkoutEffectivenessAnalyzer {

    /// Analyze workout effectiveness and generate insights
    static func generateInsights(timeSeries: [HealthMetric: MetricTimeSeries]) -> [Insight] {
        var insights: [Insight] = []

        // --- Consistency Score ---
        if let workoutSeries = timeSeries[.workoutDuration] {
            if let consistencyInsight = analyzeConsistency(workoutSeries) {
                insights.append(consistencyInsight)
            }
        }

        // --- VO2 Max Response ---
        if let vo2Series = timeSeries[.vo2Max] {
            if let vo2Insight = analyzeVO2MaxResponse(vo2Series) {
                insights.append(vo2Insight)
            }
        }

        // --- Calorie Efficiency ---
        if let calSeries = timeSeries[.activeCalories],
           let exSeries = timeSeries[.exerciseMinutes] {
            if let efficiencyInsight = analyzeCalorieEfficiency(calSeries: calSeries, exerciseSeries: exSeries) {
                insights.append(efficiencyInsight)
            }
        }

        return insights
    }

    // MARK: - Consistency

    private static func analyzeConsistency(_ workoutSeries: MetricTimeSeries) -> Insight? {
        let last28 = workoutSeries.samples(lastDays: 28)
        guard last28.count >= 7 else { return nil }

        // Count workout days per week for the last 4 weeks
        var weekWorkoutCounts: [Int] = []
        for weekOffset in 0..<4 {
            let weekStart = Date().daysAgo(28 - weekOffset * 7)
            let weekEnd = Date().daysAgo(21 - weekOffset * 7)
            let weekWorkouts = last28.filter { sample in
                sample.value > 0 && sample.date >= weekStart && sample.date < weekEnd
            }.count
            weekWorkoutCounts.append(weekWorkouts)
        }

        let weeksWithTarget = weekWorkoutCounts.filter { $0 >= 3 }.count
        let consistencyScore = Double(weeksWithTarget) / 4.0 * 100
        let weeklyAvg = Double(weekWorkoutCounts.reduce(0, +)) / 4.0

        let severity: Severity = consistencyScore < 50 ? .warning : .info
        let trend: TrendDirection = weekWorkoutCounts.last ?? 0 >= weekWorkoutCounts.first ?? 0 ? .improving : .declining

        let weekBreakdown = weekWorkoutCounts.enumerated()
            .map { "Week \($0.offset + 1): \($0.element) day\($0.element == 1 ? "" : "s")" }
            .joined(separator: ", ")

        return Insight(
            metric: .workoutDuration,
            title: Copy.Analysis.Workout.workoutConsistency,
            summary: Copy.Analysis.WorkoutEffectiveness.consistencySummary(
                consistencyPct: Int(consistencyScore),
                weeklyAvg: String(format: "%.1f", weeklyAvg),
                breakdown: weekBreakdown),
            recommendation: consistencyScore >= 75
                ? Copy.Analysis.WorkoutEffectiveness.consistencyRecHigh(
                    consistencyPct: Int(consistencyScore),
                    weeklyAvg: String(format: "%.1f", weeklyAvg),
                    breakdown: weekBreakdown)
                : Copy.Analysis.WorkoutEffectiveness.consistencyRecLow(
                    weeklyAvg: String(format: "%.1f", weeklyAvg),
                    weeksWithTarget: weeksWithTarget,
                    breakdown: weekBreakdown),
            severity: severity,
            trend: trend,
            currentValue: consistencyScore,
            baselineValue: 75,
            deviationPercent: consistencyScore - 75,
            category: .workoutEffectiveness,
            relatedMetrics: [.workoutDuration, .exerciseMinutes]
        )
    }

    // MARK: - VO2 Max Response

    private static func analyzeVO2MaxResponse(_ vo2Series: MetricTimeSeries) -> Insight? {
        let recent = vo2Series.samples(lastDays: 30)
        let older = vo2Series.samples(lastDays: 60).filter { sample in
            sample.date < Date().daysAgo(30)
        }

        guard recent.count >= 3, older.count >= 3 else { return nil }

        let recentAvg = recent.map(\.value).mean
        let olderAvg = older.map(\.value).mean
        guard olderAvg > 0 else { return nil }

        let change = ((recentAvg - olderAvg) / olderAvg) * 100
        let trend: TrendDirection = change > 2 ? .improving : (change < -2 ? .declining : .stable)

        // Weekly VO2 progression
        let allSamples = vo2Series.samples(lastDays: 30)
        var weeklyAvgs: [String] = []
        for weekOffset in stride(from: 3, through: 0, by: -1) {
            let weekStart = Date().daysAgo(7 * (weekOffset + 1))
            let weekEnd = Date().daysAgo(7 * weekOffset)
            let weekSamples = allSamples.filter { $0.date >= weekStart && $0.date < weekEnd }
            if !weekSamples.isEmpty {
                weeklyAvgs.append(String(format: "%.1f", weekSamples.map(\.value).mean))
            }
        }
        let weeklyProgression = weeklyAvgs.isEmpty ? "" : " Weekly VO2: \(weeklyAvgs.joined(separator: " \u{2192} ")) \(HealthMetric.vo2Max.unit)."

        return Insight(
            metric: .vo2Max,
            title: Copy.Analysis.Workout.vo2MaxResponse,
            summary: Copy.Analysis.WorkoutEffectiveness.vo2MaxChangeSummary(
                direction: change > 0
                    ? Copy.Analysis.WorkoutEffectiveness.directionImproved
                    : Copy.Analysis.WorkoutEffectiveness.directionDecreased,
                change: String(format: "%.1f", abs(change)),
                olderAvg: String(format: "%.1f", olderAvg),
                recentAvg: String(format: "%.1f", recentAvg),
                unit: HealthMetric.vo2Max.unit,
                weeklyProgression: weeklyProgression),
            recommendation: change > 0 ?
                "VO2 Max trending up \(String(format: "%.1f", abs(change)))% over 30 days (\(String(format: "%.1f", olderAvg)) \u{2192} \(String(format: "%.1f", recentAvg)) \(HealthMetric.vo2Max.unit)).\(weeklyProgression)" :
                "VO2 Max declined \(String(format: "%.1f", abs(change)))% over 30 days (\(String(format: "%.1f", olderAvg)) \u{2192} \(String(format: "%.1f", recentAvg)) \(HealthMetric.vo2Max.unit)).\(weeklyProgression)",
            severity: abs(change) > 5 ? .warning : .info,
            trend: trend,
            currentValue: recentAvg,
            baselineValue: olderAvg,
            deviationPercent: change,
            category: .workoutEffectiveness,
            relatedMetrics: [.vo2Max, .exerciseMinutes]
        )
    }

    // MARK: - Calorie Efficiency

    private static func analyzeCalorieEfficiency(
        calSeries: MetricTimeSeries,
        exerciseSeries: MetricTimeSeries
    ) -> Insight? {
        let aligned7d = TimeSeriesAligner.alignByDate(calSeries, exerciseSeries)
            .filter { $0.date >= Date().daysAgo(7) && $0.valueB > 0 }
        let aligned30d = TimeSeriesAligner.alignByDate(calSeries, exerciseSeries)
            .filter { $0.date >= Date().daysAgo(30) && $0.valueB > 0 }

        guard aligned7d.count >= 3, aligned30d.count >= 7 else { return nil }

        let efficiency7d = aligned7d.map { $0.valueA / $0.valueB }.mean
        let efficiency30d = aligned30d.map { $0.valueA / $0.valueB }.mean

        guard efficiency30d > 0 else { return nil }

        let change = ((efficiency7d - efficiency30d) / efficiency30d) * 100
        let trend: TrendDirection = change > 5 ? .improving : (change < -5 ? .declining : .stable)

        return Insight(
            metric: .activeCalories,
            title: Copy.Analysis.Workout.calorieEfficiency,
            summary: Copy.Analysis.WorkoutEffectiveness.efficiencySummary(
                efficiency7d: String(format: "%.1f", efficiency7d),
                efficiency30d: String(format: "%.1f", efficiency30d),
                sign: change > 0 ? "+" : "",
                change: String(format: "%.0f", change)),
            recommendation: change > 0 ?
                "Calorie efficiency up \(String(format: "%.0f", change))% this week. Burning \(String(format: "%.1f", efficiency7d)) kcal/min vs your 30-day average of \(String(format: "%.1f", efficiency30d)) kcal/min." :
                "Calorie efficiency down \(String(format: "%.0f", abs(change)))% this week. \(String(format: "%.1f", efficiency7d)) kcal/min vs your 30-day average of \(String(format: "%.1f", efficiency30d)) kcal/min.",
            severity: .info,
            trend: trend,
            currentValue: efficiency7d,
            baselineValue: efficiency30d,
            deviationPercent: change,
            category: .workoutEffectiveness,
            relatedMetrics: [.activeCalories, .exerciseMinutes]
        )
    }
}

// MARK: - InsightAnalyzer Conformance

extension WorkoutEffectivenessAnalyzer: InsightAnalyzer {
    static func generateInsights(context: AnalysisContext) -> [Insight] {
        generateInsights(timeSeries: context.timeSeries)
    }
}
